core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/HealthMonitor.scala

```diff
@@ -0,0 +1,238 @@
+    * service account.
+    */
+  private def checkGoogleGroups: Future[SubsystemStatus] = {
+    logger.debug("Checking Google Groups...")
+    // Note: call to `foldMap` depends on SubsystemStatusMonoid, defined implicitly below
+    groupsToCheck.toList.foldMap { group =>
+      googleServicesDAO.getGoogleGroup(group).map {
+        case Some(_) => OkStatus
+        case None => failedStatus(s"Could not find group: $group")
+      }
+    }
+  }
+
+  /**
+    * Checks Google bucket status by doing a Get on the token bucket using the buckets service account.
+    */
+  private def checkGoogleBuckets: Future[SubsystemStatus] = {
+    logger.debug("Checking Google Buckets...")
+    // Note: call to `foldMap` depends on SubsystemStatusMonoid, defined implicitly below
+    bucketsToCheck.toList.foldMap { bucket =>
+      googleServicesDAO.getBucket(bucket).map {
+        case Some(_) => OkStatus
+        case None => failedStatus(s"Could not find bucket: $bucket")
+      }
+    }
+  }
+
+  /**
+    * Checks Google billing status by doing a list() using the billing service account.
+    * Expects at least one account to be returned.
+    */
+  private def checkGoogleBilling: Future[SubsystemStatus] = {
+    logger.debug("Checking Google Billing...")
+    googleServicesDAO.listBillingAccountsUsingServiceCredential.map { accts =>
+      if (accts.isEmpty) failedStatus("Could not find any Rawls billing accounts")
+      else OkStatus
+    }
+  }
+
+  /**
+    * Checks Google genomics status by doing a list() using the genomics service account.
+    * Does not validate the results; only that the API call succeeds.
+    */
+  private def checkGoogleGenomics: Future[SubsystemStatus] = {
+    logger.debug("Checking Google Genomics...")
+    googleServicesDAO.listGenomicsOperations.map { _ =>
+      OkStatus
+    }
+  }
+
+  /**
+    * Checks LDAP status by doing a search and validating that we can retrieve at least one user.
+    */
+  private def checkLDAP: Future[SubsystemStatus] = {
+    logger.debug("Checking LDAP...")
+    userDirectoryDAO.listUsers.map { users =>
+      if (users.isEmpty) failedStatus("Could not find any users in LDAP")
+      else OkStatus
+    }
+  }
+
+  private def processSubsystemResult(subSystem: Subsystem, result: Future[SubsystemStatus]): Unit = {
+    result.withTimeout(futureTimeout, s"Timed out after ${futureTimeout.toString} waiting for a response from ${subSystem.toString}")
+    .recover { case NonFatal(ex) =>
+      failedStatus(ex.getMessage)
+    } map {
+      Store(subSystem, _)
+    } pipeTo self
+  }
+
+  private def store(subsystem: Subsystem, status: SubsystemStatus): Unit = {
+    data = data + (subsystem -> (status, System.currentTimeMillis))
+    logger.debug(s"New health monitor state: $data")
+  }
+
+  private def getCurrentStatus: StatusCheckResponse = {
+    val now = System.currentTimeMillis()
+    // Convert any expired statuses to unknown
```

Comment:
Hm, I was thinking about the case where a subsystem check hangs and doesn't return, but I'm enforcing a timeout on the futures so the future should fail if not complete after 1 minute. I guess this _shouldn't_ happen in practice; but it might be good as a defensive check in case our status checks stop running for any reason. (Maybe the akka scheduler dies? Not sure if that can happen.)